src/Conversion/ONNXToKrnl/Tensor/Transpose.cpp

```diff
@@ -20,83 +20,192 @@ using namespace mlir;
+    Optional<ArrayAttr> permAttr = operandAdaptor.perm();
 
+    // Input and output types.
     MemRefType inMemRefType = data.getType().cast<MemRefType>();
-    uint64_t inRank = inMemRefType.getShape().size();
-    // Convert the output type to MemRefType.
     Type outConvertedType =
         typeConverter->convertType(*op->result_type_begin());
     assert(outConvertedType && outConvertedType.isa<MemRefType>() &&
            "Failed to convert type to MemRefType");
     MemRefType outMemRefType = outConvertedType.cast<MemRefType>();
-    uint64_t outRank = outMemRefType.getShape().size();
 
     // Get shape.
     ONNXTransposeOpShapeHelper shapeHelper(op, operands, &create.krnlIE);
     shapeHelper.computeShapeAndAssertOnFailure();
+    DimsExpr &outDims = shapeHelper.getOutputDims();
+
+    // If transpose does not permute the non-1 dimensions, it is safe to lower
+    // transpose to a view op.
+    if (canBeViewOp(inMemRefType, permAttr)) {
+      Value view = create.mem.reinterpretCast(data, outDims);
+      rewriter.replaceOp(op, view);
+      return success();
+    }
+
+    // Insert an allocation and deallocation for the result of this operation.
+    Value alloc =
+        insertAllocAndDeallocSimple(rewriter, op, outMemRefType, loc, outDims);
+
+    // If the last N dimensions are not permuted, do block copying for the last
+    // N dimensions. Otherwise, do element-wise copying.
+    if (auto numLastDims = unchangedInnerDimensions(permAttr))
+      blockTranspose(data, alloc, permAttr, &create, numLastDims);
+    else
+      scalarTranspose(data, alloc, permAttr, &create);
 
-    // If the order of the dimensions whose value is not 1 does not change after
-    // transpose, it is safe to lower transpose to a view op.
+    rewriter.replaceOp(op, alloc);
+    return success();
+  }
+
+private:
+  // If transpose does not permute the non-1 dimensions, it is safe to lower
+  // transpose to a view op.
+  bool canBeViewOp(
+      MemRefType inMemRefType, Optional<ArrayAttr> permAttr) const {
     ArrayRef<int64_t> dims = inMemRefType.getShape();
+    uint64_t rank = inMemRefType.getRank();
+
     SmallVector<int64_t, 4> originalAxes;
     for (uint64_t axis = 0; axis < dims.size(); ++axis)
       if (dims[axis] != 1)
         originalAxes.emplace_back(axis);
+
     SmallVector<int64_t, 4> permutedAxes;
-    for (uint64_t i = 0; i < inRank; ++i) {
+    for (uint64_t i = 0; i < rank; ++i) {
       int64_t axis = ArrayAttrIntVal(permAttr, i);
       if (dims[axis] != 1)
         permutedAxes.emplace_back(axis);
     }
+    return (originalAxes == permutedAxes);
+  }
 
-    if (originalAxes == permutedAxes) {
-      // It is safe to lower to a view op.
-      MemRefBuilder createMemRef(rewriter, loc);
-      Value view =
-          createMemRef.reinterpretCast(data, shapeHelper.getOutputDims());
-      rewriter.replaceOp(op, view);
-      return success();
+  // Determine how many consecutive inner-most dimensions are not permuted.
```

Comment:
Is it true that this only holds if there is no map on the members? If so, I would mention it in the comments to the function.

Reply:
Yes, it's true. I need to check the map on the memrefs.

Reply:
Done. I added code to verify that there is no map on the memrefs. Mentioned that in the comment too.